crates/ra_fmt2/src/edit_tree.rs

```diff
@@ -126,21 +108,27 @@ impl Block {
         self as *const _ == other as *const _
     }
 
-    /// Returns an iterator of parents from current element.
-    fn ancestors(&self) -> Parents<'_> {
-        // what is the best way to do this Rc, Box, Cell with refs?
-        Parents( self.parent.as_ref().as_ref() )
+    /// Returns an iterator of children from current element.
+    fn children(&self) -> impl Iterator<Item=&Block> {
+        self.children.iter()
     }
 
-    /// Returns an iterator of any sibling nodes and tokens from current element.
-    fn siblings_with_tokens(&self) -> NextSibling<'_> {
-        NextSibling( self.next_sib.as_ref().as_ref() )
+    /// Traverse all blocks in order, convenience for order_flatten_blocks.
+    pub(crate) fn traverse(&self) -> impl Iterator<Item=&Block> {
+        Traversal { blocks: self.order_flatten_blocks(), idx: 0, }
     }
 
-    /// Walk all the blocks 
-    fn traverse(&self) -> WalkBlocks<'_> {
-        println!("in traverse");
-        WalkBlocks { root: self, next: Some(Branch::Continue(self)) }
+    /// Vec of all Blocks in order, parent then children.
+    fn order_flatten_blocks(&self) -> Vec<&Block> {
```

Comment:
I'm sure this could be better but hopefully this looks more like what you were thinking

Reply:
I was thinking rather about a recursive function like

```
fn traverse(block: &Block) {
    // do work
    block.children.iter().for_each(traverse)
}
```

But I think the exact logic behind traversal is not that important, as soon as it is convenient to use from withing the engine